### PR TITLE
Feature/#638 inject assets after meta charset

### PR DESCRIPTION
**Pull request by Csszabi98**

### CODE REVIEW

**Comment by Csszabi98** [comment]
@brillout Could you please reopen the related ticket? (#638)
Thanks!

File: examples/stem-react/pages/_default/Head.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ const description = 'My first Vite/Stem app'
 function Head() {
   return (
     <>
```
Comment (Csszabi98) [inline examples/stem-react/pages/_default/Head.tsx:10]: The `@brillout/stem-react` package's current implementation will always trigger the warning.
https://github.com/brillout/stem-react/blob/main/renderer/_default.page.server.tsx#L34

**Comment by brillout** [comment]
Neat 💚.

> best practice to always include it in the HTML source. Not doing so could lead to a variety of issues.

Do you know the worst-case consequence of omitting setting the charset? If the consequence is potentially bad/sneaky (hard to find that the root cause is because of a missing charset value), then we should def keep both `assertWarning()`. If the consequences are mild/obvious then I think we can show only the second warning.

**Comment by brillout** [comment]
I'm starting to question the whole idea of using RegExp to inject assets. Not only is it complex and error prone, but it also performs badly (especially with SSR that generates long HTML).

Maybe we can find a better alternative than using RegExp. Thoughts?

**Comment by Csszabi98** [comment]
Yes, I was having the same feelings during the implementation. Usually when manipulating html string source the first thing I use is DOMParser. That is more robust by miles and a much more pleasantly experience to work with as you are just using the document API. For the node environment there are two main choices in this regard, HappyDOM and JsDOM for this purpose. The former is more performant while the latter is more mature.

**Comment by brillout** [comment]
I'm thinking maybe something like this:

```js
import { html } from 'vite-plugin-ssr'

export function render() {
   return html`<html>
     <head>
       <meta charset="en" />
       ${html.ASSETS_BEGIN}
     </head>
     <body>
       ${html.ASSETS_END}
     </body>
   </html>`
}
```

That way we don't need to manipulate strings at all &ndash; we just concatenate strings instead.

**Comment by Csszabi98** [comment]
If injecting assets is the only thing required, then this seems simple enough ➕.
Only thing is that this seems like a breaking change, so I would keep the old behavior while adding this as an extra feature.

**Comment by brillout** [comment]
Since my proposal involves a little bit of architectural changes, I propse I implement this myself. Let me make some progress on the V1 design then I'll implement this on this PR. I'll keep you updated.

> I would keep the old behavior while adding this as an extra feature

Yes, that's the plan.

**Comment by Csszabi98** [comment]
Sounds great, thanks!

**Comment by brillout** [comment]
"Merged" 😀 🎉.

Many thanks for the PR, very useful to realize we should go with another path.

**Pull request merged**